c/rc_lock_supply.h

```diff
@@ -0,0 +1,217 @@
+#ifndef _RC_LOCK_SUPPLY_H_
+#define _RC_LOCK_SUPPLY_H_
+#include "ckb_syscalls.h"
+
+typedef unsigned __int128 uint128_t;
+
+enum SupplyErrorCode {
+  ERROR_EXCEED_SUPPLY = 90,
+  ERROR_SUPPLY_AMOUNT,
+};
+
+// <1 byte version> <16 bytes current supply> <16 bytes max supply> <32 bytes sUDT script hash>
+#define MIN_INFO_CELL_LEN 65
+#define INFO_CELL_CONST_OFFSET 17
+
+typedef struct SupplyContextType {
+  size_t input_info_cell_index;
+  size_t input_info_cell_count;
+  size_t output_info_cell_index;
+  size_t output_info_cell_count;
+
+  uint8_t version;
+  uint128_t input_current_supply;
+  uint128_t output_current_supply;
+  uint128_t max_supply;
+  uint8_t sudt_script_hash[32];
+
+  uint128_t input_amount;
+  uint128_t output_amount;
+} SupplyContextType;
+
+int add_assign_amount(uint128_t* sum, const uint128_t* delta) {
+  *sum += *delta;
+
+  if (*sum < *delta) {
+    return ERROR_SUPPLY_AMOUNT;
+  } else {
+    return 0;
+  }
+}
+
+int minus_assign_amount(uint128_t* sum, const uint128_t* delta) {
+  if (*sum < *delta)
+    return ERROR_SUPPLY_AMOUNT;
+  *sum -= *delta;
+  return 0;
+}
+
+typedef int (iterate_func_t)(size_t index, size_t source, SupplyContextType* ctx);
+
+int locate_info_cell(size_t index, size_t source, SupplyContextType* ctx) {
+  if (source == CKB_SOURCE_INPUT) {
+    ctx->input_info_cell_index = index;
+    ctx->input_info_cell_count += 1;
+    if (ctx->input_info_cell_count > 1) {
+      return CKB_INVALID_DATA;
+    } else {
+      return 0;
+    }
+  } else if (source == CKB_SOURCE_OUTPUT) {
+    ctx->output_info_cell_index = index;
+    ctx->output_info_cell_count += 1;
+    if (ctx->output_info_cell_count > 1) {
+      return CKB_INVALID_DATA;
+    } else {
+      return 0;
+    }
+  } else {
+    return CKB_INVALID_DATA;
+  };
+}
+
+int accumulate_amount(size_t index, size_t source, SupplyContextType* ctx) {
+  int err = 0;
+  uint128_t amount = 0;
+  uint64_t len = 16;
+  err = ckb_checked_load_cell_data((uint8_t*)&amount, &len, 0, index, source);
+  CHECK(err);
+  if (source == CKB_SOURCE_INPUT) {
+    err = add_assign_amount(&ctx->input_amount, &amount);
+    CHECK(err);
+  } else if (source == CKB_SOURCE_OUTPUT) {
+    err = add_assign_amount(&ctx->output_amount, &amount);
+    CHECK(err);
+  } else {
+    CHECK2(false, ERROR_SUPPLY_AMOUNT);
+  }
+exit:
+  return err;
+}
+
+int compare_cells_data(size_t input_index, size_t output_index) {
+  int err = 0;
+  int err2 = 0;
+  uint8_t input_version = 0;
+  uint8_t output_version = 0;
+  uint64_t input_len = 1;
+  uint64_t output_len = 1;
+  err = ckb_load_cell_data(&input_version, &input_len, 0, input_index, CKB_SOURCE_INPUT);
+  CHECK(err);
+  err = ckb_load_cell_data(&output_version, &output_len, 0, output_index, CKB_SOURCE_OUTPUT);
+  CHECK(err);
+  CHECK2(input_len == output_len, CKB_INVALID_DATA);
+  CHECK2(input_len >= MIN_INFO_CELL_LEN, CKB_INVALID_DATA);
```

Comment:
The check is duplicated with line `176` in function `check_supply()`

Reply:
I think, duplicated checking in different functions is OK.